dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java

```diff
@@ -234,49 +195,77 @@ public T singleResult(final Query query) {
-     * @return count of matching rows.
-     */
-    public int count(Criteria criteria) {
-        return ((Long) criteria.setProjection(Projections.rowCount()).uniqueResult()).intValue();
+    public int count(Context context, CriteriaQuery criteriaQuery, CriteriaBuilder criteriaBuilder, Root<T> root)
+        throws SQLException {
+        return Math.toIntExact(countLong(context, criteriaQuery, criteriaBuilder, root));
     }
 
-    /**
-     * How many rows match this query?
-     *
-     * @param query description of the rows.
-     * @return count of matching rows.
-     */
     public int count(Query query) {
-        return ((Long) query.uniqueResult()).intValue();
+        return ((Long) query.getSingleResult()).intValue();
     }
 
-    /**
-     * How many rows match these criteria?
-     *
-     * @param criteria description of the rows.
-     * @return count of matching rows.
-     */
-    public long countLong(Criteria criteria) {
-        return (Long) criteria.setProjection(Projections.rowCount()).uniqueResult();
+    public long countLong(Context context, CriteriaQuery criteriaQuery, CriteriaBuilder criteriaBuilder, Root<T>
+        root) throws SQLException {
+        Expression<Long> countExpression = criteriaBuilder.countDistinct(root);
+        criteriaQuery.select(countExpression);
+        return (Long) this.getHibernateSession(context).createQuery(criteriaQuery).getSingleResult();
+    }
+
+    public CriteriaQuery<T> getCriteriaQuery(CriteriaBuilder criteriaBuilder, Class<T> c) {
+        CriteriaQuery<T> criteriaQuery = criteriaBuilder.createQuery(c);
+        return criteriaQuery;
+    }
+
+    public CriteriaBuilder getCriteriaBuilder(Context context) throws SQLException {
+        return this.getHibernateSession(context).getCriteriaBuilder();
     }
+
+    public List<T> executeCriteriaQuery(Context context, CriteriaQuery<T> criteriaQuery, boolean cacheable, Class<T>
+        clazz, int maxResults, int offset) throws SQLException {
+        //This has to be here, otherwise a 500 gets thrown
+        Query query = this.getHibernateSession(context).createQuery(criteriaQuery);
+
+        //TODO Check if this works and is desireable
+        query.setHint("org.hibernate.cacheable", cacheable);
+//        query.setCacheable(cacheable);
+        if (maxResults != -1) {
+            query.setMaxResults(maxResults);
+        }
+        if (offset != -1) {
+            query.setFirstResult(offset);
+        }
+        return query.getResultList();
+
+    }
+
+    public List<T> findByX(Context context, Class clazz, Map<String, Object> equals, boolean cacheable, int
+        maxResults, int offset) throws SQLException {
+        CriteriaBuilder criteriaBuilder = getCriteriaBuilder(context);
+        CriteriaQuery<T> criteria = getCriteriaQuery(criteriaBuilder, clazz);
+        Root root = criteria.from(clazz);
+        criteria.select(root);
+
+        //TODO Maybe one big where, test this;;;; seems to not be necessary
```

Comment:
Not sure I understand this comment. Could we provide a better description to what is going on in this method?  Needs Javadocs at a minimum.